Transform object to yaml and back has problem

Pom as following:

            <dependency>
              <groupId>io.kubernetes</groupId>
              <artifactId>client-java</artifactId>
              <version>3.0.0</version>
            </dependency>

I am using io.kubernetes.client.util.Yaml class to do the transformation from object and back. Codes as following.

    V1Pod pod = coreV1Api.readNamespacedPod(podName, namespace, null, null, null);
    String yaml = Yaml.dump(pod);
    V1Pod pod2 = Yaml.loadAs(yaml, V1Pod.class)

I encounter the error with line 3 like following:

    org.yaml.snakeyaml.nodes.MappingNode cannot be cast to org.yaml.snakeyaml.nodes.ScalarNode
     in 'reader', line 4, column 22:
          creationTimestamp: {}
                             ^
    
      at org.yaml.snakeyaml.constructor.Constructor$ConstructMapping.constructJavaBean2ndStep(Constructor.java:270)
      at org.yaml.snakeyaml.constructor.Constructor$ConstructMapping.construct(Constructor.java:149)
      at org.yaml.snakeyaml.constructor.BaseConstructor.constructObjectNoCheck(BaseConstructor.java:207)
      at org.yaml.snakeyaml.constructor.BaseConstructor.constructObject(BaseConstructor.java:196)
      at io.kubernetes.client.util.Yaml$CustomConstructor.constructObject(Yaml.java:334)
      at org.yaml.snakeyaml.constructor.Constructor$ConstructMapping.constructJavaBean2ndStep(Constructor.java:247)
      ... 12 more
    Caused by: java.lang.ClassCastException: org.yaml.snakeyaml.nodes.MappingNode cannot be cast to org.yaml.snakeyaml.nodes.ScalarNode
      at io.kubernetes.client.util.Yaml$CustomConstructor.constructObject(Yaml.java:331)
      at org.yaml.snakeyaml.constructor.Constructor$ConstructMapping.constructJavaBean2ndStep(Constructor.java:247)
      ... 17 more
    
Here is the snippet of the yaml transformed from object

    apiVersion: v1
    kind: Pod
    metadata:
      creationTimestamp: {}
      generateName: test-776d6c86cc-
      labels:
        app: test
        app-version: '19911'
        pod-template-hash: '3328274277'
      name: test-776d6c86cc-4zwj5
      namespace: default
    ...

Here is the snippet of the yaml when I use `kubectl edit pod xxxx` command:

    apiVersion: v1
    kind: Pod
    metadata:
      creationTimestamp: "2018-12-23T01:09:18Z"
      generateName: test-776d6c86cc-
      labels:
        app: test
        app-version: "19911"
        pod-template-hash: "3328274277"
      name: test-776d6c86cc-4zwj5
      namespace: default

It seems Yaml class has some problem to convert DateTime to String

--- a/util/src/main/java/io/kubernetes/client/util/Yaml.java
+++ b/util/src/main/java/io/kubernetes/client/util/Yaml.java
@@ -358,6 +358,15 @@ public class Yaml {
       this.representers.put(IntOrString.class, new RepresentIntOrString());
       this.representers.put(byte[].class, new RepresentByteArray());
       this.representers.put(Quantity.class, new RepresentQuantity());
+      this.representers.put(DateTime.class, new RepresentDateTime());
+    }
+
+    private class RepresentDateTime implements Represent {
+      @Override
+      public Node representData(Object data) {
+        DateTime dateTime = (DateTime) data;
+        return CustomRepresenter.this.representData(data.toString());
+      }
     }
 
     private class RepresentIntOrString implements Represent {

--- a/util/src/main/java/io/kubernetes/client/util/Yaml.java
+++ b/util/src/main/java/io/kubernetes/client/util/Yaml.java
@@ -364,7 +364,6 @@ public class Yaml {
     private class RepresentDateTime implements Represent {
       @Override
       public Node representData(Object data) {
-        DateTime dateTime = (DateTime) data;
         return CustomRepresenter.this.representData(data.toString());
       }
     }

--- a/util/src/test/java/io/kubernetes/client/util/YamlTest.java
+++ b/util/src/test/java/io/kubernetes/client/util/YamlTest.java
@@ -43,6 +43,8 @@ public class YamlTest {
 
   private static final URL EXPECTED_YAML_FILE = Resources.getResource("expected.yaml");
 
+  private static final URL CREATED_TIMESTAMP_FILE = Resources.getResource("test-pod.yaml");
+
   private static final String[] kinds =
       new String[] {
         "Pod",
@@ -232,6 +234,21 @@ public class YamlTest {
           "2018-09-06T15:12:24.000Z",
           new String(pod.getMetadata().getCreationTimestamp().toString().getBytes(), UTF_8));
 
+    } catch (Exception ex) {
+      assertNull("Unexpected exception: " + ex.toString(), ex);
+    }
+  }
+
+  @Test
+  public void testDateTimeRoundTrip() {
+    // There was an issue with dumping JODA DateTime as String.
+    // This test verifies that its fixed.
+    try {
+      String data = Resources.toString(CREATED_TIMESTAMP_FILE, UTF_8);
+      V1Pod pod = Yaml.loadAs(data, V1Pod.class);
+      String output = Yaml.dump(pod);
+      V1Pod pod2 = Yaml.loadAs(output, V1Pod.class);
+      assertEquals(pod, pod2);
     } catch (Exception ex) {
       assertNull("Unexpected exception: " + ex.toString(), ex);
     }

--- a/util/src/test/resources/test-pod.yaml
+++ b/util/src/test/resources/test-pod.yaml
@@ -1,0 +1,11 @@
+apiVersion: v1
+kind: Pod
+metadata:
+  creationTimestamp: "2018-12-23T01:09:18Z"
+  generateName: test-776d6c86cc-
+  labels:
+    app: test
+    app-version: "19911"
+    pod-template-hash: "3328274277"
+  name: test-776d6c86cc-4zwj5
+  namespace: default

